File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/server/ClientCatcher.java
package lunalauf.rms.utilities.network.server;

import java.net.ServerSocket;
import java.util.concurrent.ExecutorService;

import javafx.application.Platform;
import javafx.concurrent.Worker.State;
import javafx.concurrent.WorkerStateEvent;
import javafx.event.EventHandler;

public class ClientCatcher {

    private final ClientCatchService service;

    public ClientCatcher(ExecutorService threadPool, ClientHandler clientHandler, ServerSocket serverSocket) {
        this.service = new ClientCatchService(threadPool, clientHandler, serverSocket);
        clientHandler.setOnLostClients(lostClients -> {
			Platform.runLater(this::start);
        });
    }

    public void start() {
        if (service.getState().equals(State.RUNNING))
            service.restart();
        else {
            service.reset();
            service.start();
        }
    }

    public void stop() {
        service.cancel();
    }

    public void setOnStarted(EventHandler<WorkerStateEvent> event) {
        service.setOnRunning(event);
    }

    public void setOnStopped(EventHandler<WorkerStateEvent> event) {
        service.setOnSucceeded(event);
        service.setOnCancelled(event);
        service.setOnFailed(event);
    }
}

File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/server/ClientCatcher.kt
package lunalauf.rms.utilities.network.server

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.isActive
import lunalauf.rms.utilities.network.util.Service
import java.net.ServerSocket
import java.net.SocketException
import java.net.SocketTimeoutException

class ClientCatcher(
    private val clientHandler: ClientHandler,
    private val serverSocket: ServerSocket
) : Service<Unit, Unit>(Dispatchers.IO) {
    companion object {
        private const val TIMEOUT = 1000 // ms
        private const val MAX_ITERATIONS = 30
    }

    init {
        try {
            serverSocket.setSoTimeout(TIMEOUT)
        } catch (e: SocketException) {
            logger.error("Could not set socket timeout. Behaviour unpredictable", e)
        }
        clientHandler.onLostClient { start(Unit) }
    }

    override fun CoroutineScope.run(input: Unit) {
        logger.info("Client catcher started")
        var it = 0
        while (isActive && it < MAX_ITERATIONS) {
            try {
                val socket = serverSocket.accept()
                val client = Client(socket)
                logger.info("Client accepted: {}. Passing to client handler", client.remoteAddress)
                clientHandler.handleClient(client)
            } catch (_: SocketTimeoutException) {
                it++
            }
        }
        logger.info("Client catcher stopped")
    }
}
